[dtensor] canonicalize detach callsites and use `view_as` when appropriate

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105239

This PR canonicalize the detach callsite to only call the detach
from `distribute_tensor`. Change other callsite to view_as and remove the
tensor constructor detach call

This is so that we don't detach local tensor for every op run when
rewrapping the DTensor

wanchaol committed:
[dtensor] canonicalize detach callsites

This PR canonicalize the detach callsite to only call the detach
from constructors: 1. from_local 2. distribute_tensor

This is so that we don't detach local tensor for every op run when
rewrapping the DTensor

[ghstack-poisoned]

albanD commented on torch/distributed/_tensor/api.py:
```diff
@@ -57,6 +57,9 @@ class _ToTorchTensor(torch.autograd.Function):
     @staticmethod
     def forward(ctx, input: "DTensor"):  # type: ignore[override]
         ctx.dtensor_spec = input._spec
+        # detach local tensor from its autograd graph as autograd works
```
from our offline discussion, we might want to change this to a view_as() to avoid any confusion on the autograd implication of the detach here (which is none!).

albanD commented on torch/distributed/_tensor/api.py:
```diff
@@ -111,8 +114,10 @@ def forward(  # type: ignore[override]
                     input = input.contiguous()
                     device_mesh.broadcast(input, mesh_dim=idx)
 
+        # detach the local tensor passed to DTensor since after the construction
```
Again, we should use a view here to say: we want a fresh Tensor object that shares memory and that's it.
There is no autograd concern here.

albanD commented on torch/distributed/_tensor/api.py:
```diff
@@ -459,8 +461,10 @@ def distribute_tensor(
             )
 
     assert local_tensor is not None, "distributing a tensor should not be None"
+    # detach the local tensor passed to DTensor since after the construction
```
Same as above!

wanchaol committed:
Update on "[dtensor] canonicalize detach callsites"


This PR canonicalize the detach callsite to only call the detach
from constructors: 1. from_local 2. distribute_tensor

This is so that we don't detach local tensor for every op run when
rewrapping the DTensor

[ghstack-poisoned]

albanD commented on torch/distributed/_tensor/api.py:
```diff
@@ -459,8 +460,10 @@ def distribute_tensor(
+    # of DTensor, autograd would work on top of DTensor instead of local tensor
     return DTensor(
-        local_tensor,
+        local_tensor.detach(),
```
I'm not 100% familiar with the semantic of `distribute_tensor()` but the current implementation implies that you get a DTensor with no history out of it, even if the input did require gradients. Is that accurate?

wanchaol replied:
Oh yes this is sth I would like to discuss, `distribute_tensor` is usually used to generate a DTensor and replacing it with nn.Parameters (that being said, I think `distribute_tensor` mostly being used to create leaf DTensors), so I think `detach` call is proper here? because we pass in `tensor.requires_grad` instead of `local_tensor`, it would still become leaf of the autograd graph.

I was thinking that if we create a view here, then the gradient would flow back to the original tensor (which is pre-sharding), this is sth we don't want because we want the gradient to stay sharded (`.grad` should be a sharded DTensor instead).

albanD replied:
> I was thinking that if we create a view here, then the gradient would flow back to the original tensor

It would if local_tensor ever get involved in the graph indeed.

If the semantic of this function is to create a leaf, then detach() is the right move I think yes.

wanchaol replied:
Thanks for confirming! Yeah `distribute_tensor` is used to create leaf DTensors, so let's keep the `detach` callsite here, and I removed other callsites to use `view_as` instead as we discussed :)

albanD reviewed (commented):
SGTM then!
Should we have test that ensure we create leafs here?

wanchaol committed:
Update on "[dtensor] canonicalize detach callsites and use `view_as` when appropriate"


This PR canonicalize the detach callsite to only call the detach
from constructors: 1. from_local 2. distribute_tensor

This is so that we don't detach local tensor for every op run when
rewrapping the DTensor

[ghstack-poisoned]

wanchaol commented:
> SGTM then! Should we have test that ensure we create leafs here?

sounds good! Update the PR by explicitly raise error when passing non-leaf tensor to `distribute_tensor` API and added a test.

wanchaol committed:
Update on "[dtensor] canonicalize detach callsites and use `view_as` when appropriate"


This PR canonicalize the detach callsite to only call the detach
from constructors: 1. from_local 2. distribute_tensor

This is so that we don't detach local tensor for every op run when
rewrapping the DTensor

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] canonicalize detach callsites and use `view_as` when appropriate"


This PR canonicalize the detach callsite to only call the detach
from constructors: 1. from_local 2. distribute_tensor

This is so that we don't detach local tensor for every op run when
rewrapping the DTensor

[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

wanchaol commented:
@pytorchbot merge

fduwjj commented on torch/distributed/_tensor/api.py:
```diff
@@ -57,7 +57,10 @@ class _ToTorchTensor(torch.autograd.Function):
         ctx.dtensor_spec = input._spec
-        return input._local_tensor.detach()
+        # We need to return a fresh Tensor object there as autograd metadata
+        # will be inplaced into it. So we don't want to polute the Tensor
```
nit: pollute.

fduwjj reviewed (commented):
Does this change bring perf gain for DTensor and TP?

fduwjj commented on torch/distributed/_tensor/api.py:
```diff
@@ -57,7 +57,10 @@ class _ToTorchTensor(torch.autograd.Function):
+        # We need to return a fresh Tensor object there as autograd metadata
+        # will be inplaced into it. So we don't want to polute the Tensor
+        # object stored in _local_tensor.
+        return input._local_tensor.view_as(input._local_tensor)
```
Does this mean, now the if I call `to_local` for a DTensor, we still get a leaf node?
